hash_practice/exercises.py

```diff
@@ -2,18 +2,43 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n log n)
     """
-    pass
+    letters_hash = {}
+
+    for word in strings: 
+        anagram_sorted = "".join(sorted(word)) 
+        if anagram_sorted in letters_hash: 
+            letters_hash[anagram_sorted].append(word)
+        else: 
+            letters_hash[anagram_sorted] = [word]
+
+    return list(letters_hash.values())
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    highest_k_hash = {}
+    if nums == []:
+        return nums
+
+    for num in nums:
+        if num in highest_k_hash:
+            highest_k_hash[num] += 1
+        else:
+            highest_k_hash[num] = 1
+
+    sorted_k_values = []
+    for i in range(k):
+        highest_freq_value = max(highest_k_hash, key=highest_k_hash.get)

```

Comment:
Note that you have a loop here which runs k times and inside it you're going through the dictionary (of max n elements) and finding the maximum entry.  So O(nk)